cnd/src/swap_protocols/hbit/ledger_states.rs

```diff
@@ -0,0 +1,84 @@
+use crate::swap_protocols::{
+    hbit::{LedgerState, SwapEvent},
+    LocalSwapId,
+};
+use async_trait::async_trait;
+use std::collections::HashMap;
+use tokio::sync::Mutex;
+
+#[async_trait]
+pub trait Insert: Send + Sync + 'static {
+    async fn insert(&self, key: LocalSwapId, value: LedgerState);
+}
+
+#[async_trait]
+pub trait Get: Send + Sync + 'static {
+    async fn get(&self, key: &LocalSwapId) -> Option<LedgerState>;
+}
+
+#[async_trait]
+pub trait Update: Send + Sync + 'static {
+    async fn update(&self, key: &LocalSwapId, update: SwapEvent);
+}
+
+#[derive(Default, Debug)]
+pub struct LedgerStates {
+    states: Mutex<HashMap<LocalSwapId, LedgerState>>,
+}
+
+#[async_trait]
+impl Insert for LedgerStates {
+    async fn insert(&self, key: LocalSwapId, value: LedgerState) {
+        let mut states = self.states.lock().await;
+        states.insert(key, value);
+    }
+}
+
+#[async_trait]
+impl Get for LedgerStates {
+    async fn get(&self, key: &LocalSwapId) -> Option<LedgerState> {
+        let states = self.states.lock().await;
+        states.get(key).cloned()
+    }
+}
+
+#[async_trait]
+impl Update for LedgerStates {
+    async fn update(&self, key: &LocalSwapId, event: SwapEvent) {
+        let mut states = self.states.lock().await;
+        let ledger_state = match states.get_mut(key) {
+            Some(state) => state,
+            None => {
+                tracing::warn!("Value not found for key {}", key);
+                return;
+            }
+        };
+
+        match event {
+            SwapEvent::Deployed(deployed) => ledger_state.transition_to_deployed(deployed),
+            SwapEvent::Funded(funded) => ledger_state.transition_to_funded(funded),
+            SwapEvent::Redeemed(redeemed) => {
+                // what if redeemed.secret.hash() != secret_hash in request ??
```

Comment:
Can you clarify how this can happen?

Reply:
Ahhh, funny that should come up now - this is one of those places in the code that should have the label `FIXME`. I did not write this comment and have no clue what it means, hence I'd suggest we remove the CI rule and encourage devs to use
`FIXME (tcharding): blah blah`

Reply:
> Why is this introduced now and was not raised earlier?

It is introduced now because this PR does a bunch of code duplication that uses the structs that this enum replaces.

> I'm afraid that introducing this here now means for sure that we have to refactor one or the other in the future.

We definitely have a *lot* of refactoring to do to remove all the rfc003 kruft.

> I'm not against this but we should stick to one solution and not jump back and forth.

The stated aim of this PR was to write clean code _without_ breaking anyone else's code, hence we now have multiple abstractions for the same thing. (There is also an enum in the db layer that represents this also :)

Reply:
@bonomat You commented on quite a few things that were not originally written in this PR but just copied over. I am not sure how helpful that is here :grimacing: 

We can debate how carefully one should be when copying code. I would argue quite careful but in this case, the main aim was to have a clean slate to work off from so I don't mind much :)

Reply:
How should I know this? It is green and marked as new, hence, it asked for a review ¯\_(ツ)_/¯ 

None of my comments are blocking, so don't worry too much about this.

Reply:
> How should I know this? It is green and marked as new, hence, it asked for a review ¯_(ツ)_/¯

Fair point :)